activerecord/lib/active_record/insert_all.rb

```diff
@@ -64,15 +65,25 @@ def map_key_with_value
       inserts.map do |attributes|
         attributes = attributes.stringify_keys
         attributes.merge!(scope_attributes) if scope_attributes
+        attributes.reverse_merge!(timestamps_for_create) if set_timestamps_on_create?
 
         verify_attributes(attributes)
 
-        keys.map do |key|
+        keys_including_timestamps.map do |key|
           yield key, attributes[key]
         end
       end
     end
 
+    # TODO: Consider remaining this method, as it only conditionally extends keys, not always
+    def keys_including_timestamps
+      @keys_including_timestamps ||= if set_timestamps_on_create?
+        keys + model.all_timestamp_attributes_in_model
+      else
+        keys
+      end
+    end
```

Comment:
I don't like this name, but I'm not sure of a better alternative. `extended_keys` maybe, but that's not super clear about how they're extended.